src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -196,69 +189,94 @@ private async ValueTask ProcessEventChangeAsync(ImmutableSegmentedList<bool> cha
+                    caretPosition = _dataSource.GetCaretPoint(_textView, _subjectBuffer);
+                    oldTagTrees = this.CachedTagTrees;
+                    oldState = this.State;
+
+                    textChangeRange = this.AccumulatedTextChanges;
+                    this.AccumulatedTextChanges = null;
+                }, cancellationToken).NoThrowAwaitable();
+
+                // Since we don't ever throw above, check and see if the await completed due to cancellation and do not
+                // proceed.
+                if (cancellationToken.IsCancellationRequested)
+                    return;
+
                 // if we're tagging documents that are not visible, then introduce a long delay so that we avoid
                 // consuming machine resources on work the user isn't likely to see.  ConfigureAwait(true) so that if
                 // we're on the UI thread that we stay on it.
                 //
                 // Don't do this for explicit high priority requests as the caller wants the UI updated as quickly as
                 // possible.
-                if (!highPriority)
+                if (!highPriority && !isVisible)
                 {
                     await _visibilityTracker.DelayWhileNonVisibleAsync(
                         _dataSource.ThreadingContext, _subjectBuffer, DelayTimeSpan.NonFocus, cancellationToken).ConfigureAwait(true);
                 }
 
-                await _dataSource.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-
-                _dataSource.ThreadingContext.ThrowIfNotOnUIThread();
-                if (cancellationToken.IsCancellationRequested)
-                    return;
-
                 using (Logger.LogBlock(FunctionId.Tagger_TagSource_RecomputeTags, cancellationToken))
                 {
-                    // Make a copy of all the data we need while we're on the foreground.  Then switch to a threadpool
-                    // thread to do the computation. Finally, once new tags have been computed, then we update our state
-                    // again on the foreground.
-                    var spansToTag = GetSpansAndDocumentsToTag();
-                    var caretPosition = _dataSource.GetCaretPoint(_textView, _subjectBuffer);
-                    var oldTagTrees = this.CachedTagTrees;
-                    var oldState = this.State;
-
-                    var textChangeRange = this.AccumulatedTextChanges;
-                    this.AccumulatedTextChanges = null;
-
                     // Technically not necessary since we ConfigureAwait(false) right above this.  But we want to ensure
                     // we're always moving to the threadpool here in case the above code ever changes.
                     await TaskScheduler.Default;
 
-                    cancellationToken.ThrowIfCancellationRequested();
+                    if (cancellationToken.IsCancellationRequested)
+                        return;
```

Comment:
switched to just bailing instead of throwing, to help with the debugger issue.

Reply:
note: this is def a poor-mans-fix as any cancellable call we make to any other component still throws.